apps/frontend/src/utils/ZwoParsing.ts

```diff
@@ -0,0 +1,235 @@
+import { parseXml, XmlDocument, XmlElement, XmlText } from '@rgrove/parse-xml';
+import { Workout, WorkoutPart } from '../types';
+
+export const parseZwoWorkout = (
+  str: string
+): { workout: Workout | null; errors: Set<string> } => {
+  try {
+    const document: XmlDocument = parseXml(str);
+    const workoutFileElement = document.children[0];
+    if (!isXmlElement(workoutFileElement)) {
+      return {
+        workout: null,
+        errors: new Set([`Invalid ZWO-file: Missing children-element`]),
+      };
+    }
+    const zwoWorkout = parseWorkoutFile(workoutFileElement);
+    return {
+      workout: zwoWorkout.workout
+        ? zwoWorkoutToDundringWorkout(zwoWorkout.workout)
+        : null,
+      errors: zwoWorkout.errors,
+    };
+  } catch (err) {
+    console.error(err);
+    return {
+      workout: null,
+      errors: new Set([`Invalid ZWO-file: ${err}`]),
+    };
+  }
+};
+
+type ZwoWorkout = {
+  name: string;
+  parts: Array<ZwoWorkoutPart>;
+};
+
+type ZwoWorkoutPart = SteadyState | IntervalsT | FreeRide | Warmup | Cooldown;
+
+type SteadyState = { type: 'steadystate'; duration: number; power: number };
+
+type IntervalsT = {
+  type: 'intervals';
+  repeat: number;
+  onDuration: number;
+  offDuration: number;
+  onPower: number;
+  offPower: number;
+};
+
+type FreeRide = { type: 'freeride'; duration: number };
+
+type Warmup = {
+  type: 'warmup' | 'ramp';
+  duration: number;
+  powerLow: number;
+  powerHigh: number;
+};
+
+type Cooldown = {
+  type: 'cooldown';
+  duration: number;
+  powerLow: number;
+  powerHigh: number;
+};
+
+const parseWorkoutFile = (
+  workoutFileElement: XmlElement
+): { workout: ZwoWorkout | null; errors: Set<string> } => {
+  const workoutElement = findXmlElementChildWithName(
+    workoutFileElement,
+    'workout'
+  );
+  if (!workoutElement) {
+    return {
+      workout: null,
+      errors: new Set([`Invalid ZWO-file: ${'workout-element is missing'}`]),
+    };
+  }
+
+  const workoutParts = new Array<ZwoWorkoutPart>();
+
+  const workoutPartErrors = new Set<string>();
+  workoutElement.children.flatMap((workoutPartElement) => {
+    if (!isXmlElement(workoutPartElement)) {
+      return [];
+    }
+    const parsedWorkoutPart = parseWorkoutPart(workoutPartElement);
+
+    parsedWorkoutPart.errors.forEach((error) => workoutPartErrors.add(error));
+
+    if (parsedWorkoutPart.workoutPart) {
+      workoutParts.push(parsedWorkoutPart.workoutPart);
+    }
+  });
+  const nameText = findXmlElementChildWithName(
+    workoutFileElement,
+    'name'
+  )?.children?.at(0);
+  const name = nameText instanceof XmlText ? nameText.text : 'ZWO workout';
+
+  return { workout: { parts: workoutParts, name }, errors: workoutPartErrors };
+};
```

Comment:
Not the prettiest code (whole file), but should be pretty understandable.
The Zwo-parsing is very isolated (only one exposed function), so it does not bled into the whole codebase